src/gtest/test_paymentdisclosure.cpp

```diff
@@ -0,0 +1,203 @@
+#include <gtest/gtest.h>
+
+#include "main.h"
+#include "utilmoneystr.h"
+#include "chainparams.h"
+#include "utilstrencodings.h"
+#include "zcash/Address.hpp"
+#include "wallet/wallet.h"
+#include "amount.h"
+#include <memory>
+#include <string>
+#include <set>
+#include <vector>
+#include <boost/filesystem.hpp>
+#include <iostream>
+#include "util.h"
+
+#include "paymentdisclosure.h"
+#include "paymentdisclosuredb.h"
+
+#include "sodium.h"
+
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+using namespace std;
+
+/*
+    To run tests:
+    ./zcash-gtest --gtest_filter="paymentdisclosure.*"
+
+    Note: As an experimental feature, writing your own tests may require option flags to be set.
+    mapArgs["-experimentalfeatures"] = true;
+    mapArgs["-paymentdisclosure"] = true;
+*/
+
+#define NUM_TRIES 10000
+
+#define DUMP_DATABASE_TO_STDOUT false
+
+static boost::uuids::random_generator uuidgen;
+
+static uint256 random_uint256()
+{
+    uint256 ret;
+    randombytes_buf(ret.begin(), 32);
+    return ret;
+}
+
+// Subclass of PaymentDisclosureDB to add debugging methods
+class PaymentDisclosureDBTest : public PaymentDisclosureDB {
+public:
+    PaymentDisclosureDBTest(const boost::filesystem::path& dbPath) : PaymentDisclosureDB(dbPath) {}
+
+    void DebugDumpAllStdout() {
+        assert(db != nullptr);
+        std::lock_guard<std::mutex> guard(lock_);
+
+        // Iterate over each item in the database and print them
+        leveldb::Iterator* it = db->NewIterator(leveldb::ReadOptions());
+
+        for (it->SeekToFirst(); it->Valid(); it->Next()) {
+            cout << it->key().ToString() << " : ";
+            // << it->value().ToString() << endl;
+            try {
+                std::string strValue = it->value().ToString();
+                PaymentDisclosureInfo info;
+                CDataStream ssValue(strValue.data(), strValue.data() + strValue.size(), SER_DISK, CLIENT_VERSION);
+                ssValue >> info;
+                cout << info.ToString() << std::endl;
+            } catch (const std::exception& e) {
+                cout << e.what() << std::endl;
+            }
+        }
+
+        if (false == it->status().ok()) {
+            cerr << "An error was found iterating over the database" << endl;
+            cerr << it->status().ToString() << endl;
+        }
+
+        delete it;
+    }
+};
+
+
+
+// This test creates random payment disclosure blobs and checks that they can be
+// 1. inserted and retrieved from a database
+// 2. serialized and deserialized without corruption 
+TEST(paymentdisclosure, mainnet) {
+    ECC_Start();
```

Comment:
Test failed due to this call; it can only be called once per process, and is already called in `src/gtest/test_foundersreward.cpp`. This call should be moved into `src/gtest/main.cpp`, similarly to in the Boost tests in `src/test/test_bitcoin.cpp`.

Reply:
(The call in `src/gtest/test_foundersreward.cpp` should also be removed.)

Reply:
Fixed ECC_Start

Reply:
Ok that was solving the wrong problem.  ECC_Start can be called more than once in a process, but a corresponding ECC_Stop must be called before the next invocation of ECC_Start.   It's also called in `wallet/gtest/test_wallet_zkeys.cpp`